src/lib.rs

```diff
@@ -745,256 +609,174 @@ impl BoxConstraints {
+    /// finer grained invalidation before long.
     pub fn invalidate(&mut self) {
-        self.layout_ctx.invalidate();
+        // Note: for the current functionality, we could shortcut and just
+        // request an invalidate on the window. But when we do fine-grained
+        // invalidation, we'll want to compute the invalidation region, and
+        // that needs to be propagated (with, likely, special handling for
+        // scrolling).
+        self.base_state.needs_inval = true;
     }
 
-    /// Request layout; implies invalidation.
-    pub fn request_layout(&mut self) {
-        self.layout_ctx.request_layout();
+    /// Get an object which can create text layouts.
+    pub fn text(&mut self) -> &mut Text<'b> {
+        self.win_ctx.text_factory()
     }
 
-    /// Send an event, to be handled by listeners.
-    pub fn send_event<A: Any>(&mut self, a: A) {
-        self.layout_ctx
-            .event_q
-            .push(Event::Event(self.id, Box::new(a)));
+    /// Set the cursor icon.
+    ///
+    /// Call this when handling a mouse move event, to set the cursor for the
+    /// widget. A container widget can safely call this method, then recurse
+    /// to its children, as a sequence of calls within an event propagation
+    /// only has the effect of the last one (ie no need to worry about
+    /// flashing).
+    pub fn set_cursor(&mut self, cursor: &Cursor) {
+        *self.cursor = Some(cursor.clone());
     }
 
-    /// Set or unset the widget as active.
-    // TODO: this should call SetCapture/ReleaseCapture as well.
+    /// Set the "active" state of the widget.
+    ///
+    /// The active state basically captures a mouse press inside the widget.
+    /// Thus, a button should set this to true on mouse down and false on
+    /// mouse up.
+    ///
+    /// While a widget is active, all mouse events are routed to it.
     pub fn set_active(&mut self, active: bool) {
-        self.layout_ctx.active = if active { Some(self.id) } else { None };
+        self.base_state.is_active = active;
+        // TODO: plumb mouse grab through to platform (through druid-shell)
     }
 
-    pub fn set_focused(&mut self, focused: bool) {
-        self.layout_ctx.focused = if focused { Some(self.id) } else { None };
+    /// Query the "hot" state of the widget.
+    ///
+    /// A widget is hot when the mouse is hovering.
+    pub fn is_hot(&self) -> bool {
+        self.base_state.is_hot
     }
 
-    /// Determine whether this widget is active.
+    /// Query the "active" state of the widget.
+    ///
+    /// This is the same state set by [`set_active`](#method.set_active) and
+    /// is provided as a convenience.
     pub fn is_active(&self) -> bool {
-        self.layout_ctx.active == Some(self.id)
+        self.base_state.is_active
+    }
+
+    /// Returns a reference to the current `WindowHandle`.
+    ///
+    /// Note: we're in the process of migrating towards providing functionality
```

Comment:
maybe a small tracking issue for this work?

Reply:
Done, https://github.com/xi-editor/druid/issues/95.